### PR TITLE
Eng 8075 squash subquery

### PR DESCRIPTION
**Pull request by jxchinaren**

### CODE REVIEW

File: src/frontend/org/voltdb/planner/AbstractParsedStmt.java [hunk 1]
```diff
@@ -477,7 +608,186 @@ private AbstractExpression parseOperationExpression(VoltXMLElement exprNode) {
...
+
+    /**
+     * Optimize IN expression
+     *
+     * @param inExpr
+     * @return existsExpr
+     */
+    private AbstractExpression optimizeInExpression(AbstractExpression inExpr) {
+        assert(ExpressionType.COMPARE_EQUAL == inExpr.getExpressionType());
+        if (canConvertInToExistsExpression(inExpr)) {
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/planner/AbstractParsedStmt.java]: Not a priority, but better if:

if (! canConvertInToExistsExpression(inExpr)) {
    return inExpr;
}
...

File: src/frontend/org/voltdb/planner/MaterializedViewFixInfo.java [hunk 2]
```diff
@@ -181,7 +181,9 @@ public boolean processMVBasedQueryFix(StmtTableScan mvTableScan, Set<SchemaColum
             String colName = mvCol.getName();
 
             TupleValueExpression tve = new TupleValueExpression(mvTableName, mvTableAlias, colName, colName, i);
+
             tve.setTypeSizeBytes(mvCol.getType(), mvCol.getSize(), mvCol.getInbytes());
+            tve.setOrigStmtId(mvTableScan.getStatementId());
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/planner/MaterializedViewFixInfo.java]: There are a couple of places using function "setTypeSizeBytes" to set the member information. Is that better to add one more parameter for setTypeSizeBytes and force it to be set?

For non-subquery case, the original id is -1 by default. Verified that the other places does not need to be set. So priority is low for my comment here.

File: src/frontend/org/voltdb/planner/PlanAssembler.java [hunk 3]
```diff
@@ -258,6 +257,7 @@ void setupForNewPlans(AbstractParsedStmt parsedStmt) {
                 m_parsedSelect.setHasPartitionColumnInGroupby();
             }
 
+            // FIXME: is the following scheme/comment obsolete?
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/planner/PlanAssembler.java]: Probably it should add the ticket number here : ENG-7360 for farther explanation.

**Comment by jxchinaren** [comment]
Java source code part looks generally good to me.

File: tests/frontend/org/voltdb/planner/PlannerTestCase.java [hunk 4]
```diff
@@ -39,11 +39,14 @@
     private boolean m_byDefaultInferPartitioning = true;
     private boolean m_byDefaultPlanForSinglePartition;
 
-    protected void failToCompile(String sql, String... patterns)
-    {
+    /**
+     * @param sql
+     * @return
+     */
+    private int countQuestionMarks(String sql) {
```
Comment (jxchinaren) [inline tests/frontend/org/voltdb/planner/PlannerTestCase.java]: Not sure why refactoring this into a function but not used anywhere else?

File: src/ee/common/NValue.hpp [hunk 5]
```diff
@@ -650,8 +650,20 @@ class NValue {
         return &valueChars[i];
     }
 
+    // Copy a value. If the value is inlined in a source tuple, then allocate
+    // memory from the temp string pool and copy data there
+    NValue copyNValue() const
+    {
+        NValue copy = *this;
+        if (m_sourceInlined) {
```
Comment (wolffcm) [inline src/ee/common/NValue.hpp]: This is a nice wrapper around `allocateObjectFromInlinedValue`.  I wonder if we should just be calling `copyNValue` and make `allocateObjectFromInlinedValue` private?

I need to think about this one.

File: src/ee/common/NValue.hpp [hunk 6]
```diff
@@ -650,8 +650,20 @@ class NValue {
...
+            // The NValue storage is inlined (a pointer to the backing tuple storage) and needs
+            // to be copied to a local storage
+            copy.allocateObjectFromInlinedValue(getTempStringPool());
+        }
+        return copy;
+    }
 
-  private:
+private:
```
Comment (wolffcm) [inline src/ee/common/NValue.hpp]: (Minor) this should be indented by 2 spaces to be consistent.

File: src/frontend/org/voltdb/planner/SubPlanAssembler.java [hunk 7]
```diff
@@ -1394,6 +1406,12 @@ protected static AbstractPlanNode getIndexAccessPlanForTable(StmtTableScan table
                 // Set up the similar VectorValue --> TVE replacement of the search key expression.
                 expr2 = elemExpr;
             }
+            if (expr2 instanceof AbstractSubqueryExpression) {
+                // The AbstractSubqueryExpression must be wrapped up into a
+                // ScalarValueExpression which extracts the actual row/column from
+                // the subquery
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/planner/SubPlanAssembler.java]: @alexeevm Does it mean that it supports ScalarValueExpression Index? I do not see a test case on it. Do you remember?

Comment (alexeevm) [inline src/frontend/org/voltdb/planner/SubPlanAssembler.java]: @jxchinaren Yes, it does. The test cases are in TestPlansScalarSubQueries.testWhereIndexScalar

**Pull request merged**